File: Programmers/Level 2/21.cpp
#include <string>
#include <vector>
#include <stack>
#include <cmath>
#include <algorithm>

using namespace std;
stack<int> st;

string solution(string number, int k) {
    string answer = "";
    st.push(number[0] - '0');
    int cnt = number.size() - k;
    for (int i = 1; i < number.size(); i++) {
        int num = number[i] - '0';
        while (!st.empty() && (k > 0) && st.top() < num) {
            st.pop();
            k--;
        }            
        st.push(num);
    }

    int i = 0;
    while (!st.empty()) {
        answer += to_string(st.top());
        st.pop();
    }
    reverse(answer.begin(), answer.end());
    answer = answer.substr(0, cnt);
    
    return answer;
}

File: Programmers/Level-2/21.py
def solution(number, k):
    answer = ''
    stack = [number[0]]
    
    for i in range(1, len(number)):
        while len(stack) > 0 and k > 0 and stack[-1] < number[i]:
            k -= 1
            stack.pop()
        stack.append(number[i])
    
    if k > 0:
        stack.pop()
        
    answer = ''.join(stack)
    return answer
